[Zh] sync 5 files in /security/authorization/authz-custom/ (#13151)
istio/istio.io: content/zh/docs/tasks/security/authorization/authz-ingress/index.md

## 原始客户端的源 IP 地址 {#source-ip-address-of-the-original-client}

如果数据包通过外部代理负载均衡器和 / 或 Kube-Proxy ，则客户端的原始源 IP 地址将丢失。
以下小节介绍了为不同类型的负载均衡保留原始客户端 IP 以用于日志记录或安全目的的一些策略：

1. [TCP/UDP Proxy Load Balancer](#tcp-proxy)
1. [Network Load Balancer](#network)
1. [HTTP/HTTPS Load Balancer](#http-https)

以下是 Istio 在流行的托管 Kubernetes 环境下使用 `LoadBalancer` 服务创建的负载均衡器类型，以供参考：

|Cloud Provider | Load Balancer Name            | Load Balancer Type
----------------|-------------------------------|-------------------
|AWS EKS        | Classic Elastic Load Balancer | TCP Proxy
|GCP GKE        | TCP/UDP Network Load Balancer | Network
|Azure AKS      | Azure Load Balancer           | Network
|IBM IKS/ROKS   | Network Load Balancer         | Network
|DO DOKS        | Load Balancer                 | Network
{{< tip >}}

> 您可以指示 AWS EKS 在网关服务上创建带有注释的网络负载均衡器：

Reviewer: ```suggestion
您可以指示 AWS EKS 在网关服务上创建带有注解的的 Network Load Balancer：
```

{{< tabset category-name="config-api" >}}

{{< tab name="Istio classic" category-value="istio-classic" >}}

{{< text yaml >}}
apiVersion: install.istio.io/v1alpha1
kind: IstioOperator
spec:
  meshConfig:
    accessLogEncoding: JSON
    accessLogFile: /dev/stdout
  components:
    ingressGateways:
    - enabled: true
      k8s:
        hpaSpec:
          maxReplicas: 10
          minReplicas: 5
        serviceAnnotations:
          service.beta.kubernetes.io/aws-load-balancer-type: "nlb"
{{< /text >}}

{{< /tab >}}

{{< tab name="Gateway API" category-value="gateway-api" >}}

{{< text yaml >}}
apiVersion: gateway.networking.k8s.io/v1beta1
kind: Gateway
metadata:
  name: httpbin-gateway
  annotations:
    service.beta.kubernetes.io/aws-load-balancer-type: "nlb"
spec:
  gatewayClassName: istio
  ...
{{< /text >}}

{{< /tab >}}

{{< /tabset >}}

{{< /tip >}}

### TCP/UDP 代理负载均衡器 {#tcp-proxy}

如果您使用的是 TCP/UDP 代理外部负载均衡器 (AWS Classic ELB)，
它可以使用[代理协议](https://www.haproxy.com/blog/haproxy/proxy-protocol/)

> 将原始客户端 IP 地址嵌入到分组数据中。外部负载均衡器和 Istio 入口网关都必须支持代理协议才能工作。

Reviewer: ```suggestion
将原始客户端 IP 地址嵌入到分组数据中。外部负载均衡器和 Istio Ingress 网关都必须支持代理协议才能工作。
```

  components:
    ingressGateways:
    - enabled: true
      name: istio-ingressgateway
      k8s:
        hpaSpec:
          maxReplicas: 10
          minReplicas: 5
        serviceAnnotations:
          service.beta.kubernetes.io/aws-load-balancer-proxy-protocol: "*"
        ...
{{< /text >}}

{{< /tab >}}

{{< tab name="Gateway API" category-value="gateway-api" >}}

{{< text yaml >}}
apiVersion: gateway.networking.k8s.io/v1beta1
kind: Gateway
metadata:
  name: httpbin-gateway
  annotations:
    service.beta.kubernetes.io/aws-load-balancer-proxy-protocol: "*"
spec:
  gatewayClassName: istio
  ...
---
apiVersion: autoscaling/v2
kind: HorizontalPodAutoscaler
metadata:
  name: httpbin-gateway
spec:
  scaleTargetRef:
    apiVersion: apps/v1
    kind: Deployment
    name: httpbin-gateway-istio
  minReplicas: 5
  maxReplicas: 10
{{< /text >}}

{{< /tab >}}

{{< /tabset >}}

### 网络负载均衡器 {#network}

如果您正在使用保留客户端 IP 地址的 TCP/UDP 网络负载均衡器 (AWS 网络负载均衡器、
GCP 外部网络负载均衡器、Azure 负载均衡器)，或者您正在使用轮询DNS，则您可以通过绕过
kube-proxy 并阻止其将流量发送到其他节点，使用 `externalTrafficPolicy: Local`
设置来同时保留 Kubernetes 内部的客户端 IP。

{{< warning >}}

> 对于生产部署，如果您开启了 `exteralTrafficPolicy：Local`，强烈建议 **将一个入口网关实例部署到多个节点**。否则，这将导致 **只有** 具有活动入口网关实例的节点能够接受并将传入的 NLB 流量分发到集群的其余部分，从而造成潜在的入口流量瓶颈和降低的内部负载均衡能力，甚至在具有入口网关实例的节点子集关闭时完全丧失到集群的入口流量。有关更多信息，请参阅[服务源 IP `Type=NodePort`](https://kubernetes.io/docs/tutorials/services/source-ip/#source-ip-for-services-with-type-nodeport)。

Reviewer: ```suggestion
对于生产部署，如果您开启了 `externalTrafficPolicy: Local`，强烈建议 **将一个 Ingress 网关实例部署到多个节点**。否则，这将导致 **只有** 具有活动 Ingress 网关实例的节点能够接受并将传入的 NLB 流量分发到集群的其余部分，从而造成潜在的 Ingress 流量瓶颈和降低的内部负载均衡能力，甚至在具有 Ingress 网关实例的节点子集关闭时完全丧失到集群的 Ingress 流量。有关更多信息，请参阅[服务源 IP `Type=NodePort`](https://kubernetes.io/zh-cn/docs/tutorials/services/source-ip/#source-ip-for-services-with-type-nodeport)。
```

> 对于生产部署，如果您开启了 `externalTrafficPolicy: Local`，强烈建议 **将一个 Ingress 网关实例部署到多个节点**。否则，这将导致 **只有** 具有活动 Ingress 网关实例的节点能够接受并将传入的 NLB 流量分发到集群的其余部分，从而造成潜在的 Ingress 流量瓶颈和降低的内部负载均衡能力，甚至在具有 Ingress 网关实例的节点子集关闭时完全丧失到集群的 Ingress 流量。有关更多信息，请参阅[服务源 IP `Type=NodePort`](https://kubernetes.io/zh-cn/docs/tutorials/services/source-ip/#source-ip-for-services-with-type-nodeport)。

Reviewer 2: ```suggestion
对于生产部署，如果您开启了 `externalTrafficPolicy: Local`，
强烈建议 **将一个 Ingress 网关实例部署到多个节点**。
否则，这将导致 **只有** 具有活动 Ingress 网关实例的节点能够接受并将传入的 NLB 流量分发到集群的其余部分，
从而造成潜在的 Ingress 流量瓶颈和降低的内部负载均衡能力，
甚至在具有 Ingress 网关实例的节点子集关闭时完全丧失到集群的 Ingress 流量。
有关更多信息，请参阅[服务源 IP `Type=NodePort`](https://kubernetes.io/zh-cn/docs/tutorials/services/source-ip/#source-ip-for-services-with-type-nodeport)。
```

{{< /warning >}}

> 使用以下命令更新入口网关以设置 `exteralTrafficPolicy：Local` 以保留入口网关上的原始客户端源IP:

Reviewer: ```suggestion
使用以下命令更新 Ingress 网关以设置 `exteralTrafficPolicy：Local` 以保留 Ingress 网关上的原始客户端源 IP:
```

  components:
    ingressGateways:
    - enabled: true
      name: istio-ingressgateway
      k8s:
        hpaSpec:
          maxReplicas: 10
          minReplicas: 5
        serviceAnnotations:
          service.beta.kubernetes.io/aws-load-balancer-proxy-protocol: "*"
        ...
{{< /text >}}

{{< /tab >}}

{{< tab name="Gateway API" category-value="gateway-api" >}}

{{< text yaml >}}
apiVersion: gateway.networking.k8s.io/v1beta1
kind: Gateway
metadata:
  name: httpbin-gateway
  annotations:
    service.beta.kubernetes.io/aws-load-balancer-proxy-protocol: "*"
spec:
  gatewayClassName: istio
  ...
---
apiVersion: autoscaling/v2
kind: HorizontalPodAutoscaler
metadata:
  name: httpbin-gateway
spec:
  scaleTargetRef:
    apiVersion: apps/v1
    kind: Deployment
    name: httpbin-gateway-istio
  minReplicas: 5
  maxReplicas: 10
{{< /text >}}

{{< /tab >}}

{{< /tabset >}}

### 网络负载均衡器 {#network}

如果您正在使用保留客户端 IP 地址的 TCP/UDP 网络负载均衡器 (AWS 网络负载均衡器、
GCP 外部网络负载均衡器、Azure 负载均衡器)，或者您正在使用轮询DNS，则您可以通过绕过
kube-proxy 并阻止其将流量发送到其他节点，使用 `externalTrafficPolicy: Local`
设置来同时保留 Kubernetes 内部的客户端 IP。

{{< warning >}}

> For production deployments it is strongly recommended to **deploy an ingress gateway pod to multiple nodes** if you enable `externalTrafficPolicy: Local`. Otherwise, this creates a situation where **only** nodes with an active ingress gateway pod will be able to accept and distribute incoming NLB traffic to the rest of the cluster, creating potential ingress traffic bottlenecks and reduced internal load balancing capability, or even complete loss of ingress traffic to the cluster if the subset of nodes with ingress gateway pods go down. See [Source IP for Services with `Type=NodePort`](https://kubernetes.io/docs/tutorials/services/source-ip/#source-ip-for-services-with-type-nodeport) for more information.

Reviewer: Line 302 is required to localize too.

{{< /warning >}}

Update the ingress gateway to set `externalTrafficPolicy: Local` to preserve the

> original client source IP on the ingress gateway using the following command:

Reviewer: Lines 305-306 are required to localize too.

  meshConfig:
    accessLogEncoding: JSON
    accessLogFile: /dev/stdout
    defaultConfig:
      gatewayTopology:
        numTrustedProxies: 1
{{< /text >}}

## IP-based allow list and deny list

**何时使用 `ipBlocks` 与 `emoteIpBlocks`:** 如果您使用 X-Forwarded-For HTTP 头部
或代理协议来确定原始客户端 IP 地址，则应在 `AuthorizationPolicy` 中使用 `emoteIpBlocks`。
如果您使用的是 `ExtraalTrafficPolicy：Local`，那么您的 `AuthorizationPolicy` 中应该使用
`ipBlocks`。

> |Load Balancer Type |Source of Client IP   | `ipBlocks` vs. `remoteIpBlocks`

Reviewer: ```suggestion
| 负载均衡器类型 | 客户端源 IP   | `ipBlocks` 与 `remoteIpBlocks`
```
原文是正确的，无需修改

  meshConfig:
    accessLogEncoding: JSON
    accessLogFile: /dev/stdout
    defaultConfig:
      gatewayTopology:
        numTrustedProxies: 1
{{< /text >}}

## IP-based allow list and deny list

**何时使用 `ipBlocks` 与 `emoteIpBlocks`:** 如果您使用 X-Forwarded-For HTTP 头部
或代理协议来确定原始客户端 IP 地址，则应在 `AuthorizationPolicy` 中使用 `emoteIpBlocks`。
如果您使用的是 `ExtraalTrafficPolicy：Local`，那么您的 `AuthorizationPolicy` 中应该使用
`ipBlocks`。

> |Load Balancer Type |Source of Client IP   | `ipBlocks` vs. `remoteIpBlocks`

Reviewer: Line 354 is required to localize too.

Author: Here, I don't see a problem.

192.168.10.15
{{< /text >}}

***remoteIpBlocks:***

{{< text bash >}}
$ CLIENT_IP=$(kubectl get pods -n istio-system -o name -l istio=ingressgateway | sed 's|pod/||' | while read -r pod; do kubectl logs "$pod" -n istio-system | grep remoteIP; done | tail -1 | awk -F, '{print $4}' | awk -F: '{print $2}' | sed 's/ //') && echo "$CLIENT_IP"
192.168.10.15
{{< /text >}}

{{< /tab >}}

{{< tab name="Gateway API" category-value="gateway-api" >}}

***ipBlocks:***

{{< text bash >}}
$ CLIENT_IP=$(kubectl get pods -n foo -o name -l istio.io/gateway-name=httpbin-gateway | sed 's|pod/||' | while read -r pod; do kubectl logs "$pod" -n foo | grep remoteIP; done | tail -1 | awk -F, '{print $3}' | awk -F: '{print $2}' | sed 's/ //') && echo "$CLIENT_IP"
192.168.10.15
{{< /text >}}

***remoteIpBlocks:***

{{< text bash >}}
$ CLIENT_IP=$(kubectl get pods -n foo -o name -l istio.io/gateway-name=httpbin-gateway | sed 's|pod/||' | while read -r pod; do kubectl logs "$pod" -n foo | grep remoteIP; done | tail -1 | awk -F, '{print $4}' | awk -F: '{print $2}' | sed 's/ //') && echo "$CLIENT_IP"
192.168.10.15
{{< /text >}}

{{< /tab >}}

{{< /tabset >}}

> * Update the `ingress-policy` to include your client IP address:

Reviewer: Line 510 is required to localize too.
